sklego/meta/grouped_transformer.py

```diff
@@ -0,0 +1,153 @@
+import numpy as np
+import pandas as pd
+from sklearn.base import BaseEstimator, TransformerMixin, clone
+from sklearn.utils import check_array
+from sklearn.utils.validation import check_is_fitted
+
+from sklego.common import as_list
+
+
+class GroupedTransformer(BaseEstimator, TransformerMixin):
+    """
+    Construct a transformer per data group. Splits data by groups from single or multiple columns
+    and transforms remaining columns using the transformers corresponding to the groups.
+
+    :param transformer: the transformer to be applied per group
+    :param groups: the column(s) of the matrix/dataframe to select as a grouping parameter set
+    :param use_global_model: Whether or not to fall back to a general transformation in case a group
+                             is not found during `.transform()`
+    """
+
+    def __init__(self, transformer, groups=0, use_global_model=True):
+        self.transformer = transformer
+        self.groups = groups
+        self.use_global_model = use_global_model
+
+    def __check_value_columns(self, X):
+        """Do basic checks on the value columns"""
+        if isinstance(X, pd.DataFrame):
+            X_value = check_array(X.drop(columns=self.groups))
+        else:
+            # If not a df, we impose standard checks also on grouping columns
+            X = check_array(X)
+            X_value = np.delete(X, as_list(self.groups), axis=1)
+
+        return X_value
+
+    def __check_grouping_columns(self, X):
+        """Do basic checks on grouping columns"""
+        missings = False
+
+        if isinstance(X, pd.DataFrame):
+            X_group = X.loc[:, as_list(self.groups)]
+            if X_group.isnull().sum().sum() > 0:
+                missings = True
+        else:
+            X_group = check_array(X)[:, as_list(self.groups)]
+            if np.isnan(X_group).sum(axis=None) > 0:
+                missings = True
+
+        if missings:
+            raise ValueError("X has NaN values")
+
+        # The grouping part we always want as a DataFrame with range index
+        return pd.DataFrame(X_group).reset_index(drop=True)
```

Comment:
Are these methods the exact same as in the `GroupedEstimator`? Since the methods are static, they might as well be functions. This might be nicer with regards to both unit testing as well as less code.

Reply:
Unfortunately not. I think that the setup in `GroupedTransformer` is way cleaner and easier than the one in `GroupedEstimator` (from the start splitting the data in a grouper-part and a values-part vs keeping everything together). The check on the values part is just a regular `check_array`. 

The biggest difference is that shrinkage can add an extra column to the grouper-part. 

I would say opening a new issue to make both in the 'new style' would be better than doing it in this PR, do you agree?

Reply:
Let's keep it as is for now then. That grouped estimator needs a big refactor at some point anyway, I'm fine if that happens in a different PR.